lib/backend/kubernetes/kubernetes.go

```diff
@@ -0,0 +1,263 @@
+/*
+Copyright 2022 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kubernetes
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"os"
+	"strings"
+	"sync"
+
+	"github.com/gravitational/teleport/lib/backend"
+	kubeutils "github.com/gravitational/teleport/lib/kube/utils"
+	"github.com/gravitational/trace"
+	corev1 "k8s.io/api/core/v1"
+	kubeerrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	applyconfigv1 "k8s.io/client-go/applyconfigurations/core/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	secretIdentifierName   = "state"
+	namespaceEnv           = "KUBE_NAMESPACE"
+	teleportReplicaNameEnv = "TELEPORT_REPLICA_NAME"
+	releaseNameEnv         = "RELEASE_NAME"
+)
+
+// InKubeCluster detemines if the agent is running inside a Kubernetes cluster and has access to
+// service account token and cluster CA. Besides, it also validates the presence of `KUBE_NAMESPACE`
+// and `TELEPORT_REPLICA_NAME` environment variables to generate the secret name.
+func InKubeCluster() bool {
+	_, _, err := kubeutils.GetKubeClient("")
+
+	return err == nil &&
+		len(os.Getenv(namespaceEnv)) > 0 &&
+		len(os.Getenv(teleportReplicaNameEnv)) > 0
+}
+
+// Backend uses Kubernetes Secrets to store identities.
+type Backend struct {
+	// kubernetes client
+	k8sClientSet *kubernetes.Clientset
+	namespace    string
+	secretName   string
+	replicaName  string
+
+	// Mutex is used to limit the number of concurrent operations per agent to 1 so we do not need
+	// to handle retries locally.
+	// The same happens with SQlite backend.
+	mu *sync.Mutex
+}
+
+// New returns a new instance of Kubernetes Secret identity backend storage.
+func New() (*Backend, error) {
+
+	restClient, _, err := kubeutils.GetKubeClient("")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return &Backend{
+		k8sClientSet: restClient,
+		namespace:    os.Getenv(namespaceEnv),
+		replicaName:  os.Getenv(teleportReplicaNameEnv),
+		secretName: fmt.Sprintf(
+			"%s-%s",
+			os.Getenv(teleportReplicaNameEnv),
```

Comment:
Should we check these `Getenv`s or is an empty string valid for all of them?

Reply:
The empty string is already checked when `InKubeCluster` is called and we can assume that we are already safe here

Reply:
I think we should double check them here anyway
Maybe just call the method above: `InKubeCluster`
Someone might use this public method assuming it's all that is needed to start a new backend